### PR TITLE
FNN

### LINKED ISSUE #40: Connect learner knn from package FNN
[CRAN](https://cran.r-project.org/web/packages/FNN/index.html)

### PR DESCRIPTION
**Pull request by be-marc**
Closes mlr-org/mlr3learners#40

### CODE REVIEW

File: R/LearnerClassifFNN.R [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      model = self$model
+      newdata = task$data(cols = task$feature_names)
+
+      if (self$predict_type == "response") {
+        p = invoke(FNN::knn, train = model$data, test = newdata, cl = model$target[[1]], .args = model$pars)
+        PredictionClassif$new(task = task, response = p)
+      } else {
+        p = invoke(FNN::knn, train = model$data, test = newdata, cl = model$target[[1]], prob = TRUE, .args = model$pars)
+
+        if (task$properties != "twoclass") {
```
Comment (mllg) [inline R/LearnerClassifFNN.R:48]: The check should be before the model fit.

File: R/LearnerClassifFNN.R [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#'
+#' @description
+#' A [mlr3::LearnerClassif] for a classification fnn implemented in [FNN::knn()] in package \CRANpkg{FNN}.
+#'
+#' @export
+LearnerClassifFNN <- R6Class("LearnerClassifFNN",
+  inherit = LearnerClassif,
+  public = list(
+    initialize = function() {
+      ps <- ParamSet$new(
```
Comment (mllg) [inline R/LearnerClassifFNN.R]: <-

File: R/LearnerClassifFNN.R [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        predict_types = c("response", "prob"),
+        param_set = ps,
+        properties = c("twoclass", "multiclass")
+      )
+    },
+
+    train_internal = function(task) {
+      list(
+        target = task$data(cols = task$target_names),
+        data = task$data(cols = task$feature_names),
```
Comment (mllg) [inline R/LearnerClassifFNN.R]: A bit nit picky, but I think we could do a single "query" here for the data.

File: R/LearnerClassifFNN.R [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        p = invoke(FNN::knn, train = model$data, test = newdata, cl = model$target[[1]], .args = model$pars)
+        PredictionClassif$new(task = task, response = p)
+      } else {
+        p = invoke(FNN::knn, train = model$data, test = newdata, cl = model$target[[1]], prob = TRUE, .args = model$pars)
+
+        if (task$properties != "twoclass") {
+          stop("Probabilities are not available for multiclass")
+        }
+
+        # Predicted probabilities refer to the winning class
```
Comment (mllg) [inline R/LearnerClassifFNN.R:53]: AFAICT the code is right, but the term "winning" class confuses me.

File: R/LearnerRegrFNN.R [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          ParamInt$new(id = "k", default = 1, lower = 1L, tags = "train"),
+          ParamFct$new(id = "algorithm", default = "kd_tree", levels = c("kd_tree", "cover_tree", "brute"), tags = "train")
+        )
+      )
+
+      super$initialize(
+        id = "regr.fnn",
+        packages = "FNN",
+        feature_types = c("integer", "numeric"),
+        predict_types = c("response"),
```
Comment (mllg) [inline R/LearnerRegrFNN.R]: c()

File: R/LearnerRegrFNN.R [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#'
+#' @description
+#' A [mlr3::LearnerRegr] for a regression fnn implemented in [FNN::knn()] in package \CRANpkg{FNN}.
+#'
+#' @export
+LearnerRegrFNN <- R6Class("LearnerRegrFNN",
+  inherit = LearnerRegr,
+  public = list(
+    initialize = function() {
+      ps <- ParamSet$new(
```
Comment (mllg) [inline R/LearnerRegrFNN.R]: <-

**Review by mllg — CHANGES_REQUESTED** [review]

File: R/LearnerClassifFNN.R (hunk 4 above)
Comment (be-marc) [reply R/LearnerClassifFNN.R:53]: They used  this term in the `FNN` manual.  Maybe 'Predicted probability refers to the proportion of votes for the predicted class'?

**Review by mllg — APPROVED** [review]

**Pull request merged**